---
title: "Peak visualization using a genome browser"
author: "Heather Wick, Upendra Bhattarai, Meeta Mistry"
date: "Aug 16th, 2024"
---

Contributors: Heather Wick, Upendra Bhattarai, Meeta Mistry

Approximate time: 

## Learning Objectives

* Visualize peaks in IGV (Integrative Genomics Viewer).
* Understand the file formats used for visualiztion.
* Use R package (graphViz/rtracklayer) to plot and study differential peak enrichment between experimental groups.


## Overview

Integrate Genomics Viewer (IGV) is an open-source desktop genome visualization tool. Below we will learn about the file formats used to visualize peaks in IGV and use it to explore peaks in our experiment samples. We will also use graphViz/rtracklayer to plot the peak signals in R. 

Use material from variant calling workshop (setting up IGV)[https://hbctraining.github.io/variant_analysis/lessons/12_IGV.html]

***

*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
